File: Topic Wise/Two Pointers/threeSum.py
def threeSum(nums):
    nums.sort()
    res=[]
    n=len(nums)

    for i in range(n-2):

        if i>0 and nums[i]==nums[i-1]:
            continue

        l,r=i+1,n-1  # noqa: E741
        
        while l<r:
            t=nums[i]+nums[l]+nums[r]

            if t<0:
                l+=1  # noqa: E741
            elif t>0:
                r-=1
            else:
                res.append((nums[i],nums[l],nums[r]))

                while l<r and nums[l]==nums[l+1]:
                    l+=1  # noqa: E741
                while l<r and nums[r]==nums[r-1]:
                    r-=1
                
                l+=1  # noqa: E741
                r-=1
    
    return set(res)

a=[-1,0,1,2,-1,-4]
print(threeSum(list(a))) # {(-1, -1, 2), (-1, 0, 1)}

File: LC Mothly/2024/FEB/threeSum.java
// Given an integer array nums,return all the triplets[nums[i],nums[j],nums[k]]such that i!=j,i!=k,and j!=k,and nums[i]+nums[j]+nums[k]==0.

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class threeSum {
    public static void main(String[] args) {

        int[] arr = { -1, 0, 1, 2, -1, -4 };
        int n = arr.length;
        List<List<Integer>> ans = findtriplet(n, arr);
        for (List<Integer> it : ans) {
            System.out.print("[");
            for (Integer i : it) {
                System.out.print(i + " ");
            }
            System.out.print("] ");
        }
        System.out.println();
    }

    public static List<List<Integer>> triplet(int n, int[] nums) {
        // Brute force approach O(N^3)
        Set<List<Integer>> st = new HashSet<>();
        // int n = nums.length;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                for (int k = j + 1; k < n; k++) {
                    if (nums[i] + nums[j] + nums[k] == 0) {
                        List<Integer> temp = Arrays.asList(nums[i], nums[j], nums[k]);
                        temp.sort(null);
                        st.add(temp);
                    }
                }
            }
        }
        List<List<Integer>> ans = new ArrayList<>(st);
        return ans;
    }
    
    // better approach - O(N^2)
    public static List<List<Integer>> findtriplet(int n, int[] nums) {
        Set<List<Integer>> st = new HashSet<>();
        for (int i = 0; i < n; i++) {
            Set<Integer> hashset = new HashSet<>();
            for (int j = i + 1; j < n; j++) {
                int third = -(nums[i] + nums[j]);
                if (hashset.contains(third)) {
                    List<Integer> temp = Arrays.asList(nums[i], nums[j], third);
                    temp.sort(null);
                    st.add(temp);
                }
                hashset.add(nums[j]);
            }
        }
        List<List<Integer>> ans = new ArrayList<>(st);
        return ans;
    }
}
